pkg/graveler/ref/manager.go

```diff
@@ -412,19 +445,72 @@ func (m *Manager) ListCommits(ctx context.Context, repositoryID graveler.Reposit
 }
 
 func (m *Manager) FillGenerations(ctx context.Context, repositoryID graveler.RepositoryID) error {
-	_, err := m.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		return tx.Exec(`WITH RECURSIVE cte AS
-								(
-									SELECT id, 1 AS generation,creation_date
-									FROM graveler_commits
-									WHERE repository_id = $1 AND parents IS NULL OR array_length(parents,1) = 0
-									UNION ALL
-									SELECT DISTINCT(c.id), cte.generation+1 AS generation,c.creation_date
-									FROM graveler_commits c INNER JOIN cte
-									ON cte.id = ANY(c.parents) WHERE repository_id = $1
-								)
-								UPDATE graveler_commits u SET generation = cte.generation FROM cte WHERE u.id = cte.id;`,
-			repositoryID)
+	iter, err := m.ListCommits(ctx, repositoryID)
+	if err != nil {
+		return err
+	}
+	defer iter.Close()
+
+	commitIDToNode := make(map[graveler.CommitID]CommitNode)
+	var rootCommitID graveler.CommitID
+
+	// make a commits "tree" from map data structure
+	for iter.Next() {
+		commit := iter.Value()
+		if len(commit.Parents) == 0 {
```

Comment:
can we possibly have more than one commit without any parents? @arielshaqed @talSofer